### PR TITLE
Add tests for IndexMapping Tool

### PR DESCRIPTION
**Pull request by dbwiddis**
### Description

Adds test configuration for IndexMapping Tool

Updates test configuration for CatIndex Tool

### Check List
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/tests/api/cat.test.ts [hunk 1]
```diff
@@ -6,9 +6,20 @@
...
 
 const provider = ApiProviderFactory.create();
-const runner = new (class CatIndicesRunner extends QARunner {})(provider);
+
+const runner = new (class CatIndexRunner extends QARunner {
+    protected async beforeAll(clusterStateId: string): Promise<void> {
+      if (clusterStateId !== 'spider') {
+        throw new Error('unexpected cluster state id');
+      }
+      await OpenSearchTestIndices.delete('spider');
```
Comment (joshuali925) [inline src/tests/api/cat.test.ts:18]: no need to delete? since it won't change and it takes a while to create

Comment (dbwiddis) [reply src/tests/api/cat.test.ts]: Sure, was mostly copying from what you had earlier :)

**Review by sejli — APPROVED** [review]

**Comment by dbwiddis** [comment]
Agent config used for IndexMappingTool below.

```json
{
  "name": "tool-register-agent",
  "description": "Flow template",
  "use_case": "REGISTER_AGENT",
  "version": {
    "template": "1.0.0",
    "compatibility": [
      "2.12.0",
      "3.0.0"
    ]
  },
  "workflows": {
    "provision": {
      "user_params": {},
      "nodes": [
        {
          "id": "create_connector_1",
          "type": "create_connector",
          "previous_node_inputs": {},
          "user_inputs": {
            "version": "1",
            "name": "Claude instant runtime Connector",
            "protocol": "aws_sigv4",
            "description": "The connector to BedRock service for claude model",
            "actions": [
              {
                "headers": {
                  "x-amz-content-sha256": "required",
                  "content-type": "application/json"
                },
                "method": "POST",
                "request_body": "{\"prompt\":\"${parameters.prompt}\", \"max_tokens_to_sample\":${parameters.max_tokens_to_sample}, \"temperature\":${parameters.temperature},  \"anthropic_version\":\"${parameters.anthropic_version}\" }",
                "action_type": "predict",
                "url": "https://bedrock-runtime.us-west-2.amazonaws.com/model/anthropic.claude-instant-v1/invoke"
              }
            ],
            "credential": {
              "access_key": "REDACTED",
              "secret_key": "REDACTED"
            },
            "parameters": {
              "endpoint": "bedrock-runtime.us-west-2.amazonaws.com",
              "content_type": "application/json",
              "auth": "Sig_V4",
              "max_tokens_to_sample": "8000",
              "service_name": "bedrock",
              "temperature": "0.0001",
              "response_filter": "$.completion",
              "region": "us-west-2",
              "anthropic_version": "bedrock-2023-05-31"
            }
          }
        },
        {
          "id": "create_t2ppl_connector",
          "type": "create_connector",
          "previous_node_inputs": {},
          "user_inputs": {
            "version": "1",
            "name": "sagemaker: t2ppl",
            "protocol": "aws_sigv4",
            "description": "The connector to BedRock service for claude model",
            "actions": [
              {
                "action_type": "predict",
                "method": "POST",
                "headers": {
                  "content-type": "application/json"
                },
                "url": "https://runtime.sagemaker.${parameters.region}.amazonaws.com/endpoints/t2ppl-mix-1225/invocations",
                "request_body": "{\"prompt\":\"${parameters.prompt}\"}"
              }
            ],
            "credential": {
              "access_key": "REDACTED",
              "secret_key": "REDACTED"
            },
            "parameters": {
              "region": "ap-northeast-1",
              "service_name": "sagemaker"
            }
          }
        },
        {
          "id": "register_model_2",
          "type": "register_remote_model",
          "previous_node_inputs": {
            "create_connector_1": "connector_id"
          },
          "user_inputs": {
            "description": "test model",
            "deploy": true,
            "name": "claude-instant"
          }
        },
        {
          "id": "register_sparse_model_1",
          "type": "register_local_custom_model",
          "user_inputs": {
            "name": "neural-sparse/opensearch-neural-sparse-tokenizer-v1",
            "version": "1.0.0",
            "description": "This is a neural sparse tokenizer model: It tokenize input sentence into tokens and assign pre-defined weight from IDF to each. It serves only in query.",
            "function_name": "SPARSE_TOKENIZE",
            "model_format": "TORCH_SCRIPT",
            "model_content_size_in_bytes": 567691,
            "model_content_hash_value": "b3487da9c58ac90541b720f3b367084f271d280c7f3bdc3e6d9c9a269fb31950",
            "created_time": 1696913667239,
            "model_type": "sparse",
            "embedding_dimension": "1",
            "framework_type": "sentence_transformers",
            "url": "https://artifacts.opensearch.org/models/ml-models/amazon/neural-sparse/opensearch-neural-sparse-tokenizer-v1/1.0.0/torch_script/opensearch-neural-sparse-tokenizer-v1-1.0.0.zip"
          }
        },
        {
          "id": "register_t2ppl_model",
          "type": "register_remote_model",
          "previous_node_inputs": {
            "create_t2ppl_connector": "connector_id"
          },
          "user_inputs": {
            "description": "t2ppl fine-tuned model",
            "deploy": true,
            "name": "t2ppl fine-tuned model"
          }
        },
        {
          "id": "deploy_sparse_model_1",
          "type": "deploy_model",
          "previous_node_inputs": {
            "register_sparse_model_1": "model_id"
          }
        },
        {
          "id": "cat_index_tool",
          "type": "create_tool",
          "previous_node_inputs": {},
          "user_inputs": {
            "type": "CatIndexTool",
            "name": "CatIndexTool",
            "description": "Use this tool to get OpenSearch index information: (health, status, index, uuid, primary count, replica count, docs.count, docs.deleted, store.size, primary.store.size).",
            "parameters": {
              "index": ".kibana"
            }
          }
        },
        {
          "id": "index_mapping_tool",
          "type": "create_tool",
          "previous_node_inputs": {},
          "user_inputs": {
            "type": "IndexMappingTool",
            "name": "IndexMappingTool",
            "description": "Use this tool to get OpenSearch index mapping information.",
            "parameters": {}
          }
        },
        {
          "id": "visualization_tool",
          "type": "create_tool",
          "previous_node_inputs": {},
          "user_inputs": {
            "description": "Use this tool to find user created visualizations. This tool takes the visualization name as input and returns the first 3 matching visualizations",
            "include_output_in_agent_response": true,
            "type": "VisualizationTool",
            "parameters": {
              "index": ".kibana"
            },
            "name": "VisualizationTool"
          }
        },
        {
          "id": "TransferQuestionToPPLAndExecuteTool",
          "type": "create_tool",
          "previous_node_inputs": {
            "register_t2ppl_model": "model_id"
          },
          "user_inputs": {
            "type": "PPLTool",
            "name": "TransferQuestionToPPLAndExecuteTool",
            "description": "Use this tool to transfer natural language to generate PPL and execute PPL to query inside. Use this tool after you know the index name, otherwise, call IndexRoutingTool first. The input parameters are: 'index' for the index name, and 'question' for UserQuestion.",
            "parameters": {
              "response_filter": "$.completion",
              "prompt": "Below is an instruction that describes a task, paired with the index and corresponding fields that provides further context. Write a response that appropriately completes the request.\n\n### Instruction:\nI have an opensearch index with fields in the following. Now I have a question: ${indexInfo.question} Can you help me generate a PPL for that?\n\n### Index:\n${indexInfo.indexName}\n\n### Fields:\n${indexInfo.mappingInfo}\n\n### Response:\n"
            },
            "include_output_in_agent_response": true
          }
        },
        {
          "id": "search_alerts_tool",
          "type": "create_tool",
          "user_inputs": {
            "type": "SearchAlertsTool",
            "name": "SearchAlertsTool",
            "parameters": {}
          }
        },
        {
          "id": "search_monitors_tool",
          "type": "create_tool",
          "user_inputs": {
            "type": "SearchMonitorsTool",
            "name": "SearchMonitorsTool",
            "parameters": {}
          }
        },
        {
          "id": "search_anomoly_detectors_tool",
          "type": "create_tool",
          "user_inputs": {
            "type": "SearchAnomalyDetectorsTool",
            "name": "SearchAnomalyDetectorsTool",
            "parameters": {}
          }
        },
        {
          "id": "search_anomoly_results_tool",
          "type": "create_tool",
          "user_inputs": {
            "type": "SearchAnomalyResultsTool",
            "name": "SearchAnomalyResultsTool",
            "parameters": {}
          }
        },
        {
          "id": "vector_db_tool",
          "type": "create_tool",
          "user_inputs": {
            "type": "VectorDBTool",
            "name": "VectorDBTool",
            "descriptions": "This is a tool that performs knn-based dense retrieval. It takes 1 argument named input which is a string query for dense retrieval. The tool returns the dense retrieval results for the query.",
            "parameters": {
              "model_id": "ksNnFo0BY4jgIz2mWrHU",
              "index": "my_test_data",
              "embedding_field": "embedding",
              "source_field": "[\"text\"]",
              "input": "${parameters.question}"
            }
          }
        },
        {
          "id": "rag_tool",
          "type": "create_tool",
          "user_inputs": {
            "type": "RAGTool",
            "name": "RAGTool",
            "parameters": {
              "inference_model_id": "${{ register_model_2.model_id }}",
              "embedding_model_id": "ksNnFo0BY4jgIz2mWrHU",
              "index": "my_test_data",
              "embedding_field": "embedding",
              "source_field": "[\"text\"]",
              "input": "${parameters.question}",
              "prompt": "\n\nHuman:\" turn\" You are a professional data analysist. You will always answer question based on the given context first. If the answer is not directly shown in the context, you will analyze the data and find the answer. If you don't know the answer, just say don't know. \n\n Context:\n${parameters.output_field}\n\nHuman:${parameters.input}\n\nAssistant:"
            }
          }
        },
        {
          "id": "neural_sparse_knowledge_base_tool",
          "type": "create_tool",
          "previous_node_inputs": {
            "deploy_sparse_model_1": "model_id"
          },
          "user_inputs": {
            "name": "OpensearchKnowledgeBaseTool",
            "type": "NeuralSparseSearchTool",
            "description": "A tool to search the Opensearch knowledge base, the knowledge base consists of segments of OpenSearch documents. You should always search data with this tool when encountering general questions about Opensearch. But for questions about current concerete cluster, use this tool can not help you. If this tool provides useful info, give the answer and also quote the origin doc. If this tool can not provide knowledge you need, give answer based on your own knowledge. Action Input: <natrual language keywords for question>",
            "parameters": {
              "index": "knowledge_base",
              "embedding_field": "sparse_embedding",
              "source_field": "[\"title\",\"body\"]",
              "doc_size": "10",
              "input": "${parameters.question}"
            }
          }
        },
        {
          "id": "sub_agent",
          "type": "register_agent",
          "previous_node_inputs": {
            "index_mapping_tool": "tools",
            "register_model_2": "model_id"
          },
          "user_inputs": {
            "parameters": {},
            "app_type": "chatbot",
            "name": "Sub Agent",
            "description": "this is a test agent",
            "llm.parameters": {
              "max_iteration": "5",
              "stop_when_no_tool_found": "true",
              "response_filter": "$.completion"
            },
            "memory": {
              "type": "conversation_index"
            },
            "type": "conversational"
          }
        },
        {
          "id": "agent_tool",
          "type": "create_tool",
          "previous_node_inputs": {
            "sub_agent": "agent_id"
          },
          "user_inputs": {
            "description": "Agent Tool",
            "include_output_in_agent_response": true,
            "type": "AgentTool",
            "parameters": {
              "max_iteration": "5"
            },
            "name": "AgentTool"
          }
        },
        {
          "id": "ml_model_tool",
          "type": "create_tool",
          "previous_node_inputs": {
            "register_model_2": "model_id"
          },
          "user_inputs": {
            "parameters": {
              "prompt": "\n\nHuman:\" turn\" You are an AI that only speaks JSON. Do not write normal text. Output should follow example JSON format: \n\n {\"response\": [\"question1\", \"question2\"]}\n\n. \n\nHuman:\" turn\":You will be given a chat history between OpenSearch Assistant and a Human.\nUse the context provided to generate follow up questions the Human would ask to the Assistant.\nThe Assistant can answer general questions about logs, traces and metrics.\nAssistant can access a set of tools listed below to answer questions given by the Human:\nQuestion suggestions generator tool\nHere's the chat history between the human and the Assistant.\n${parameters.AgentTool.output}\nUse the following steps to generate follow up questions Human may ask after the response of the Assistant:\nStep 1. Use the chat history to understand what human is trying to search and explore.\nStep 2. Understand what capabilities the assistant has with the set of tools it has access to.\nStep 3. Use the above context and generate follow up questions.Step4:You are an AI that only speaks JSON. Do not write normal text. Output should follow example JSON format: \n\n {\"response\": [\"question1\", \"question2\"]} \n \n----------------\n\nAssistant:"
            },
            "description": "A general tool to answer any question.",
            "alias": "language_model_tool",
            "include_output_in_agent_response": true,
            "name": "QuestionSuggestor",
            "type": "MLModelTool"
          }
        },
        {
          "id": "root_agent",
          "type": "register_agent",
          "previous_node_inputs": {
            "agent_tool": "tools",
            "register_model_2": "model_id",
            "ml_model_tool": "tools"
          },
          "user_inputs": {
            "parameters": {
              "prompt": "Answer the question as best you can."
            },
            "app_type": "chatbot",
            "name": "Root agent",
            "description": "this is the root agent",
            "tools_order": [
              "agent_tool",
              "ml_model_tool"
            ],
            "memory": {
              "type": "conversation_index"
            },
            "type": "flow"
          }
        }
      ],
      "edges": [
        {
          "source": "register_model_2",
          "dest": "rag_tool"
        }
      ]
    }
  }
}
```

Similar config used for Cat Indices Tool, the only difference being list of tools in previous_node_inputs for sub_agent
```json
  "id": "sub_agent",
  "type": "register_agent",
  "previous_node_inputs": {
    "cat_index_tool": "tools",
    "register_model_2": "model_id"
  },
```

**Comment by joshuali925** [comment]
@dbwiddis i see you have `ml_model_tool` in the root agent, there is also a sub agent, which agent should the tests run against? if the root agent, why is `ml_model_tool` needed? if sub agent, why is root agent needed?

```json
            "tools_order": [
              "agent_tool",
              "ml_model_tool"
            ],
```

**Pull request merged**